### PR TITLE
feat(ci): implement domain-relevance gating in analyze_code_changes

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

This PR implements Phase 1 Domain-Relevance Gating in `.github/workflows/ci_pipeline.yml` (`analyze_code_changes`), reducing queued hardware jobs by up to 78% on domain-specific pull requests while preserving 100% test coverage.

### Why this change is being made & problem being solved
Previously, `analyze_code_changes` triggered all 17+ CI test jobs unconditionally for any `.py` file change. On PRs that only modified post-training scripts, inference helpers, GPU configs, or Pathways runners, running unrelated hardware suites consumed 4–7 TPU/GPU workers unnecessarily, causing severe queue congestion on self-hosted runners. Furthermore, because `tpu-tests` and `cpu-tests` were single combined matrix jobs, pre-training unit and integration tests could not be skipped on post-training pull requests.

### Specific implementation details
1. Added 4 domain-relevance boolean outputs (`run_pretrain_tests`, `run_posttrain_tests`, `run_pathways_tests`, `run_gpu_tests`) to `analyze_code_changes`.
2. Split `tpu-tests` and `cpu-tests` into dedicated pre-training (`tpu-pretrain-tests`, `cpu-pretrain-tests`) and post-training (`tpu-posttrain-tests`, `cpu-posttrain-tests`) matrix jobs so they can be independently gated by `run_pretrain_tests` and `run_posttrain_tests`.
3. Gated `gpu-tests`, `maxtext_tpu_pathways_unit_tests`, and `maxtext_tpu_pathways_integration_tests` on their respective domain flags (`run_gpu_tests == 'true'`, `run_pathways_tests == 'true'`).
4. Reorganized intelligent path-matching rules in Bash:
   - **GPU only** (`src/maxtext/configs/gpu/`, `src/maxtext/inference/gpu/`, `tests/end_to_end/gpu/`): Runs `gpu-tests` and notebooks, skipping all TPU and Pathways jobs.
   - **Pathways only** (`src/maxtext/inference/jetstream_pathways/`, `run_pathways_tests.yml`): Runs `pathways-tests` only, skipping all TPU, CPU, and GPU jobs.
   - **Post-Training only** (`src/maxtext/trainers/post_train/`, `tests/post_training/`): Skips `gpu-tests`, `pathways-tests`, `tpu-pretrain-tests`, and `cpu-pretrain-tests` (saving 7 test jobs / 78% reduction).
   - **General Inference only** (`src/maxtext/inference/`, `tests/inference/`): Runs `tpu-pretrain-tests` and `cpu-pretrain-tests`, skipping `gpu-tests` and `pathways-tests`.
   - **Notebooks only** (`.ipynb` / `.md`): Enables `run_notebooks=true` while skipping all heavy TPU/GPU hardware test matrices.
   - **Core Modeling / Shared Code / Configs**: Fails open (`set_all_flags "true"`), running 100% of all test suites unconditionally.

### Comparison: `origin/main` vs. PR #4723 Domain Gating

| Modified PR Scope | TPU Pretrain (`unit`, `int`) | TPU Posttrain (`post-unit`) | GPU Tests (`unit`, `int`) | Pathways Tests (`unit`, `int`) | Notebooks | Job Savings vs. `origin/main` |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| **GPU only** (`configs/gpu/`, `inference/gpu/`, `end_to_end/gpu/`) | ❌ Skipped | ❌ Skipped | ✅ Runs | ❌ Skipped | ✅ Runs | **7 test jobs saved (TPU & Pathways skipped)** |
| **Pathways only** (`jetstream_pathways/`, `run_pathways_tests.yml`) | ❌ Skipped | ❌ Skipped | ❌ Skipped | ✅ Runs | ❌ Skipped | **7 test jobs saved (TPU, CPU & GPU skipped)** |
| **Post-Training only** (`post_train/`, `tests/post_training/`) | ❌ **Skipped** *(was ✅ in main)* | ✅ Runs | ❌ **Skipped** *(wasted job in main: ran 0 posttrain tests)* | ❌ **Skipped** *(wasted job in main: ran 0 posttrain tests)* | ✅ Runs | **7 test jobs saved (78% reduction)** |
| **Inference only** (`inference/`, `tests/inference/`) | ✅ Runs | ✅ Runs | ❌ **Skipped** *(wasted job in main: pretrain only)* | ❌ **Skipped** *(wasted job in main: pretrain only)* | ❌ Skipped | **4 heavy hardware jobs saved** |
| **Notebooks only** (`.ipynb` / `.md`) | ❌ Skipped | ❌ Skipped | ❌ Skipped | ❌ Skipped | ✅ Runs | **7 test jobs saved (100% of test matrices)** |
| **Documentation only** (`*.md`) | ❌ Skipped | ❌ Skipped | ❌ Skipped | ❌ Skipped | ❌ Skipped | **100% of all CI worker jobs skipped** |
| **Core Modeling / Configs / Shared Code** | ✅ Runs | ✅ Runs | ✅ Runs | ✅ Runs | ✅ Runs | **0 jobs skipped (Fail-open safety across 100% of suites)** |

# Tests

- Verified YAML syntax and schema formatting:
  ```bash
  pre-commit run --files .github/workflows/ci_pipeline.yml
  ```
- Tested Bash regex matching across GPU-only, Pathways-only, post-training, inference, notebook, doc-only, and core dependency change sets.
- Verified fail-open fallback behavior for shared modeling layer edits.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: .github/workflows/ci_pipeline.yml [hunk 1]
```diff
@@ -58,6 +58,11 @@ jobs:
     outputs:
       run_tests: ${{ steps.check.outputs.run_tests }}
       run_notebooks: ${{ steps.check.outputs.run_notebooks }}
+      run_pretrain_tests: ${{ steps.check.outputs.run_pretrain_tests }}
+      run_posttrain_tests: ${{ steps.check.outputs.run_posttrain_tests }}
+      run_pathways_tests: ${{ steps.check.outputs.run_pathways_tests }}
+      run_inference_tests: ${{ steps.check.outputs.run_inference_tests }}
```
Comment (shralex) [inline .github/workflows/ci_pipeline.yml]: where is run_inference_tests used ?

**Review by shralex — COMMENTED** [review]
Should we also check gpu and pathways-only changes? 

# Check if ONLY GPU files were changed
if echo "$CHANGED_FILES" | grep -E '(src/maxtext/configs/gpu/|src/maxtext/inference/gpu/|tests/end_to_end/gpu/)' > /dev/null && \
   ! echo "$CHANGED_FILES" | grep -v -E '(src/maxtext/configs/gpu/|src/maxtext/inference/gpu/|tests/end_to_end/gpu/|\.md$)' > /dev/null; then
  echo "Only GPU files changed, enabling GPU tests only."
  set_all_flags "false"
  echo "run_gpu_tests=true" >> $GITHUB_OUTPUT
  echo "run_notebooks=true" >> $GITHUB_OUTPUT
  exit 0
fi

# Check if ONLY Pathways files were changed
if echo "$CHANGED_FILES" | grep -E '(jetstream_pathways|run_pathways_tests\.yml)' > /dev/null && \
   ! echo "$CHANGED_FILES" | grep -v -E '(jetstream_pathways|run_pathways_tests\.yml|\.md$)' > /dev/null; then
  echo "Only Pathways files changed, enabling Pathways tests only."
  set_all_flags "false"
  echo "run_pathways_tests=true" >> $GITHUB_OUTPUT
  exit 0
fi

**Review by shralex — APPROVED** [review]

**Review by YixuanWang-99 — APPROVED** [review]

File: .github/workflows/ci_pipeline.yml (hunk 1 above)
Comment (RexBearIU) [reply .github/workflows/ci_pipeline.yml]: Because inference tests currently execute as part of the TPU and CPU unit and integration suites, I updated the general inference rule to enable run_pretrain_tests=true and removed the redundant run_inference_tests output flag.

**Comment by RexBearIU** [comment]
Thank you for the suggestions @shralex. I added explicit gating rules for both GPU only and Pathways only changes, which now skip all unrelated TPU, CPU, or GPU runner queues on those PRs. I also streamlined the routing script with helper functions for cleaner maintenance.

File: .github/workflows/ci_pipeline.yml [hunk 2]
```diff
@@ -82,57 +112,74 @@ jobs:
...
-          echo "run_tests=true" >> $GITHUB_OUTPUT
-          echo "run_notebooks=true" >> $GITHUB_OUTPUT
-
-          # 1. Check if only documentation files (.md) were changed
+          # Sequence of checks:
+          # - Non-PR / Diff error: Run all suites (fail-open)
+          # - Docs only: Skip all tests and notebooks
+          # - Notebooks only: Run notebooks only
+          # - Core CI & dependencies: Run all suites
+          # - GPU only: Run GPU tests and notebooks only
```
Comment (SurbhiJainUSC) [inline .github/workflows/ci_pipeline.yml]: We should not run notebook tests for GPU changes as all the notebook tests are for TPUs.

File: .github/workflows/ci_pipeline.yml [hunk 3]
```diff
@@ -67,10 +71,36 @@ jobs:
       - name: Check for Code Changes
         id: check
         run: |
+          # Helper to set all flags to 'true' or 'false'
```
Comment (SurbhiJainUSC) [inline .github/workflows/ci_pipeline.yml]: This inline bash script is getting quite large and makes it hard to read. Can we extract this logic into a standalone bash script like `.github/scripts/analyze_code_changes.sh`? It would make the YAML cleaner and allow us to test the gating logic locally without having to push a commit to Github Actions.

File: .github/workflows/ci_pipeline.yml (hunk 2 above)
Comment (RexBearIU) [reply .github/workflows/ci_pipeline.yml]: Good catch, I updated the GPU only gating rule to enable GPU tests only and skip notebook tests.

File: .github/workflows/ci_pipeline.yml (hunk 3 above)
Comment (RexBearIU) [reply .github/workflows/ci_pipeline.yml]: That's great suggestion! Extracted the routing logic into .github/scripts/analyze_code_changes.sh. This keeps the workflow file clean and makes it easy to run and test gating rules locally.

**Review by SurbhiJainUSC — APPROVED** [review]
